### ISSUE TITLE
High level construct for an AppSync Elasticsearch data source

### ISSUE THREAD

**Issue by bartcallant** [aws/aws-cdk#6063]
While working on an application I needed to setup an AppSync API which accesses Elasticsearch.

### Use Case
Currently this has to be done through the `CfnXyz` construct. Implementing a higher level construct would simplify the usage of Elasticsearch as a data source.

### Proposed Solution
A simple class to create an Elasticsearch data source would do. Implementation could be done similar to the existing `DynamoDbDataSource` and `LambdaDataSource`. 

Implementation could be as simple as this:

```ts
/**
 * Properties for an AppSync Elasticsearch data source
 */
export interface ElasticsearchDataSourceProps extends BaseDataSourceProps {
    /**
     * Region for the Amazon Elasticsearch Service domain
     */
    readonly region: string;
    /**
     * Endpoint for the Amazon Elasticsearch Service domain
     */
    readonly endpoint: string;
}

/**
 * An AppSync data source backed by Elasticsearch
 */
export class ElasticsearchDataSource extends BaseDataSource {
    constructor(scope: Construct, id: string, props: ElasticsearchDataSourceProps) {
        super(scope, id, props, {
            type: 'AMAZON_ELASTICSEARCH',
            elasticsearchConfig: {
                awsRegion: props.region,
                endpoint: props.endpoint
            }
        });
    }
}
```

### Other
There was a [PR](https://github.com/aws/aws-cdk/pull/4438) which already added a base class for a data source> This makes it easy to add a new resolver.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by hoegertn** [comment]
This looks great. Just go ahead and open a PR with your code. Please also add some basic tests for the new data source.

**Comment by bartcallant** [comment]
I've extended the integration test. There I noticed some mapping templates are also included in the code. So I've looked into the console and listed them below here:

# Request mapping template

<details>
    <summary>Get document by id</summary>

    ```vtl
    #**
        Use the 'path' to get a single value by id.
        You must replace the <index> and <type> placeholder with actual values.
    *#
    {
        "version": "2017-02-28",
        "operation": "GET",
        "path": "/<index>/<type>/${context.arguments.id}",
        "params": {}
    }
    ```
</details>

<details>
    <summary>Simple term query</summary>

    ```vtl
    #**
        The 'params' key accepts any valid Elasticsearch DSL expression.
        You must replace the <index> and <type> placeholder with actual values.
    *#
    {
        "version":"2017-02-28",
        "operation":"GET",
        "path":"/<index>/<type>/_search",
        "params":{
            "body": {
                "from": 0,
                "size": 50,
                "query": {
                    "term" :{
                        "<field>":"${context.arguments.<field>}"
                    }
                }
            }
        }
    }
    ```
</details>

<details>
    <summary>Paginate with fixed-size pages</summary>

    ```vtl
    #**
        The 'params' key accepts any valid Elasticsearch DSL expression.
        You must replace the <index> and <type> placeholder with actual values.
    *#
    {
        "version": "2017-02-28",
        "operation": "GET",
        "path": "/<index>/<type>/_search",
        "params": {
            "body": {
                "from": ${context.arguments.from},
                "size": ${context.arguments.size}
            }
        }
    }
    ```
</details>

<details>
    <summary>Get all documents within a 20 mile radius</summary>

    ```vtl
    #**
        The 'params' key accepts any valid Elasticsearch DSL expression.
        You must replace the <index> and <type> placeholder with actual values.
    *#
    {
        "version":"2017-02-28",
        "operation":"GET",
        "path":"/<index>/<type>/_search",
        "params":{
            "body": {
                "query": {
                  "filtered" : {
                      "query" : {
                          "match_all" : {}
                      },
                      "filter" : {
                          "geo_distance" : {
                              "distance" : "20mi",
                              "location" : {
                                  "lat" : 47.6205,
                                  "lon" : 122.3493
                              }
                          }
                      }
                  }
                }
            }
        }
    }
    ```
</details>

# Response mapping templates

<details>
    <summary>Return single document's source</summary>

    ```vtl
    #**
        $context.result contains the full response of the Elasticsearch query.
        You can use this template to return the data from an Elasticsearch
        query that returns a single result.
    *#
    $util.toJson($context.result.get("_source"))
    ```
</details>

<details>
    <summary>Return a list of document sources</summary>

    ```
    #**
        $context.result contains the full response of the Elasticsearch query.
        Select a subset of information or iterate through hits to return the
        same shape as is expected by this field.
    *#
    [
        #foreach($entry in $context.result.hits.hits)
            ## $velocityCount starts at 1 and increments with the #foreach loop **
            #if( $velocityCount > 1 ) , #end
            $util.toJson($entry.get("_source"))
        #end
    ]
    ```
</details>

Which of these should I include? The response templates seem good to include. For the request templates, only `Get all documents within a 20 mile radius` seems quite to specific to include. What do you think?

Changes can already be viewed [here](https://github.com/bartcallant/aws-cdk/commit/3275cdb7b4df783dd7202b577cfcccd52f0dc6a6)

**Comment by adamelmore** [comment]
@bartcallant @BryanPan342 where does this effort stand? Can I be helpful in getting it across the finish line?

**Comment by BryanPan342** [comment]
@adamelmore This [PR](https://github.com/aws/aws-cdk/pull/6108) has a couple of comments to how to expand on this PR to bring it home. Mainly just adding an interface in `@aws-cdk/aws-elasticsearch` for abstraction for the `addElasticsearchDataSource` function. Similiar to the `addLambdaDataSource` and `addDynamoDbDataSource` with their `IFunction` and `ITable` params.

**Comment by adamelmore** [comment]
@BryanPan342 if I wanted to pick up the baton, would it be kosher for me to open a new PR carrying the author's work forward and addressing the review feedback in #6108? I have a need for this and would love to help out if I won't be stepping on any toes.

**Comment by BryanPan342** [comment]
@adamelmore yeah go for it :)

**Comment by adamelmore** [comment]
FYI, going to wait on the L2 construct for Elasticsearch which is [currently in progress](https://github.com/aws/aws-cdk/pull/8369). Will quickly follow-up with a corresponding AppSync data source PR once that's merged.

**Comment by cyberwombat** [comment]
Hi @adamelmore - looks like the L2 may be ready? Would love this ES data source!

**Comment by adamelmore** [comment]
> Hi @adamelmore - looks like the L2 may be ready? Would love this ES data source!

Thanks for the reminder! Will get back to this shortly.

**Comment by tsykora-verimatrix** [comment]
Any ETA on the L2 ElasticSearch data source availability, please? We've lost days trying to go the CfnDataSource / CfnResolver route, but that seems to have its own issues (type_name lookup issues). Thank you.

**Comment by ignaciolarranaga** [comment]
Guys, make it work with the following (based on the original example):
```
import { Construct } from '@aws-cdk/core';
import { Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { GraphqlApi as GraphqlApiOriginal, BaseDataSourceProps, BackedDataSource } from '@aws-cdk/aws-appsync';
import { Domain } from '@aws-cdk/aws-elasticsearch';

/**
 * Properties for an AppSync Elasticsearch data source
 */
export interface ElasticsearchDataSourceProps extends BaseDataSourceProps {
  /**
   * Region for the Amazon Elasticsearch Service domain
   */
  readonly region: string;
}

/**
* An AppSync data source backed by Elasticsearch
*/
export class ElasticsearchDataSource extends BackedDataSource {
  constructor(scope: Construct, id: string, domain: Domain, props: ElasticsearchDataSourceProps) {
    super(scope, id, props, {
      type: 'AMAZON_ELASTICSEARCH',
      elasticsearchConfig: {
        awsRegion: props.region,
        endpoint: `https://${domain.domainEndpoint}`,
      }
    });

    // @ts-ignore
    domain.grantRead(this);
    // @ts-ignore
    domain.grantWrite(this);

    this.serviceRole!.attachInlinePolicy(
      new Policy(this, 'HTTPAccess', {
        statements: [
          new PolicyStatement({
            actions: [
              "es:ESHttpDelete",
              "es:ESHttpHead",
              "es:ESHttpGet",
              "es:ESHttpPost",
              "es:ESHttpPut"
            ],
            resources: [domain.domainArn],
          })
        ]
      })
     )
  }
}

export class GraphqlApi extends GraphqlApiOriginal {
  addElasticSearchDataSource(id: string, domain: Domain, options: ElasticsearchDataSourceProps) {
    return new ElasticsearchDataSource(this, id, domain, {
      api: this,
      name: options === null || options === void 0 ? void 0 : options.name,
      description: options === null || options === void 0 ? void 0 : options.description,
      region: options.region,
    });
  }
}
```

Hope it helps.

**Comment by tombuckley91** [comment]
@adamelmore - Hi Adam, did you get anywhere with this? There's currently no one assigned to this so it seems like it's gone stale. Thanks for your work on this anyway 👍

**Comment by adamelmore** [comment]
> @adamelmore - Hi Adam, did you get anywhere with this? There's currently no one assigned to this so it seems like it's gone stale. Thanks for your work on this anyway 👍

Sorry all, I fell off the wagon here after wrapping up the Elasticsearch L2 work! With my current priorities, I don't see me getting to this soon, unfortunately.

**Comment by tombuckley91** [comment]
@adamelmore - That's too bad, but thanks for the reply, and good luck with your current work!

**Comment by BryanPan342** [comment]
I created a PR for this issue :) #14651

I'm not an expert Elasticsearch user, so I would love to get a pass from people who use it.
